### PR TITLE
Refine schema error reason message

### PR DESCRIPTION
**Pull request by ori-shalom**
This PR adds tests to the behavior that was added in #735.
The purpose of the tests is to verify that the validated value doesn't appear in the `SchemaError.Reason` field.

Additionally, this PR refines the messages in `SchemaError.Reason` and makes sure `Reason` is always set.

### CODE REVIEW

File: openapi3/issue735_test.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+			value:            map[string]interface{}{"foo": 42},
+			extraNotContains: []interface{}{42},
+		},
+		{
+			name:             "additional properties other schema type",
+			schema:           NewSchema().WithAdditionalProperties(NewStringSchema()),
+			value:            map[string]interface{}{"foo": 42},
+			extraNotContains: []interface{}{42},
+		},
+		// TODO: uncomment when https://github.com/getkin/kin-openapi/pull/747 is merged
```
Comment (fenollp) [inline openapi3/issue735_test.go:142]: It's merged

File: openapi3/issue735_test.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+			}),
+			value:            map[string]interface{}{"foo": 42},
+			extraNotContains: []interface{}{42},
+		},
+		// TODO: uncomment when https://github.com/getkin/kin-openapi/issues/502 is fixed
+		//{
+		//	name: "read only properties",
+		//	schema: NewSchema().WithProperties(map[string]*Schema{
+		//		"foo": {ReadOnly: true},
+		//	}).WithNoAdditionalProperties(),
```
Comment (fenollp) [inline openapi3/issue735_test.go:164]: We should rename this method to Without... in another PR

File: openapi3/schema.go [hunk 3]
```diff
@@ -784,6 +784,11 @@ func (schema *Schema) WithAnyAdditionalProperties() *Schema {
 	return schema
 }
 
+func (schema *Schema) WithNoAdditionalProperties() *Schema {
```
Comment (fenollp) [inline openapi3/schema.go:787]: Well actually in this PR

File: openapi3/schema.go [hunk 4]
```diff
@@ -1915,15 +1930,20 @@ func (schema *Schema) visitJSONObject(settings *schemaValidationSettings, value
...
 }
 
-func (schema *Schema) expectedType(settings *schemaValidationSettings, typ string) error {
+func (schema *Schema) expectedType(settings *schemaValidationSettings, value interface{}) error {
 	if settings.failfast {
 		return errSchema
 	}
+
+	a := "a"
+	if strings.Trim(schema.Type[:1], "aeiou") == "" {
```
Comment (fenollp) [inline openapi3/schema.go:1939]: Please just compare to the few possible values explicitly with a switch statement.

**Review by fenollp — CHANGES_REQUESTED** [review]
LGTM. A few things to fix and we should look into that readOnly fields issue

**Comment by ori-shalom** [comment]
Thanks for the review.
I fixed your comments + rebased and uncommented the `additionalProperties` test.
As for the `readOnly` & `writeOnly` issue, I think we can address in another PR.

**Review by fenollp — APPROVED** [review]

**Pull request merged**